_tools/types/main.go

```diff
@@ -0,0 +1,333 @@
+			return err
+		}
+		analyzeCode(driver, uastTypes)
+	}
+
+	formatMarkdownTable(drivers, uastTypes)
+	return nil
+}
+
+type driverStats struct {
+	url          string
+	language     string
+	path         string
+	skip         bool
+	fixturesNum  int
+	fixturesUast map[string]int
+	codeUast     map[string]int
+}
+
+// listDrivers lists all available drivers.
+func listDrivers() ([]*driverStats, error) {
+	fmt.Fprintf(os.Stderr, "discovering all available drivers\n")
+	langs, err := discovery.OfficialDrivers(context.TODO(), &discovery.Options{
+		NoStatic: true,
+	})
+	if err != nil {
+		return nil, err
+	}
+	drivers := make([]*driverStats, 0, len(langs))
+	for _, l := range langs {
+		drivers = append(drivers, &driverStats{
+			language:     l.Language,
+			url:          l.RepositoryURL(),
+			path:         l.RepositoryURL()[strings.LastIndex(l.RepositoryURL(), "/"):],
+			fixturesUast: make(map[string]int),
+			codeUast:     make(map[string]int),
+		})
+	}
+	fmt.Fprintf(os.Stderr, "%d drivers available on-line\n", len(langs))
+	return drivers, nil
+}
+
+// maybeCloneOrPullAll either clones repos to path in local FS or, if already preset,
+// pulls the latest master for each of them.
+func maybeCloneOrPullAll(drivers []*driverStats) error {
+	fmt.Fprintf(os.Stderr, "cloning %d drivers to %s\n", len(drivers), reposRootPath)
+	err := os.MkdirAll(reposRootPath, os.ModePerm)
+	if err != nil {
+		return err
+	}
+
+	var (
+		cr = make(chan int, 3)
+		wg sync.WaitGroup
+	)
+	for i := range drivers {
+		wg.Add(1)
+		go maybeCloneOrPullAsync(drivers[i], &wg, cr)
+	}
+	wg.Wait()
+	return nil
+}
+
+func maybeCloneOrPullAsync(d *driverStats, wg *sync.WaitGroup, concurent chan int) error {
+	defer wg.Done()
+
+	concurent <- 1
+	defer func() {
+		<-concurent
+	}()
+
+	return maybeCloneOrPull(d)
+}
+
+func maybeCloneOrPull(d *driverStats) error {
+	repoPath := path.Join(reposRootPath, d.path)
+	_, err := os.Stat(repoPath)
+	if err != nil && !os.IsNotExist(err) {
+		fmt.Fprintf(os.Stderr, "error: %s\n", err)
+		return err
+	}
+
+	if os.IsNotExist(err) {
+		fmt.Fprintf(os.Stderr, "%s does not exist, cloning from %s\n", repoPath, d.url)
+		cmd := exec.Command("git", "clone", d.url+".git")
+		cmd.Dir = reposRootPath
+		err = cmd.Run()
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	fmt.Fprintf(os.Stderr, "%s dir exists, will git pull instead\n", repoPath)
+	cmd := exec.Command("git", "pull", "origin", "master")
+	cmd.Dir = repoPath
+	err = cmd.Run()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type uastType struct {
+	name string
+}
+
+// isUsedIn checks if current UAST type is used in the given package.
+func (u *uastType) isUsedIn() {
+	// TODO
+}
+
+// findUastTypesInSdk finds all types from SDK.
+func findUastTypesInSdk() []uastType {
+	var out []uastType
+	// TODO: load package, iterate all structs
+	types := []interface{}{
+		uast.Position{},
+		uast.Positions{},
+		// only embeding GenNode?
```

Comment:
Sorry! This is meant to be the question about criteria to select types from UAST package: would it be useful to limit report only to the things that embed `GenNode` or just all the types found in that package? 

That would be a useful for further automation of type selection process, in case we want to do some filtering.

Reply:
I see what you mean now. I think we should list all the types including positions. We could omit those because we can't possibly miss them in new drivers, but those are only two types anyway, so it makes sense to include them for completeness.